File: 2552-maximum-sum-of-distinct-subarrays-with-length-k/maximum-sum-of-distinct-subarrays-with-length-k.cpp
class Solution {
public:
    long long maximumSubarraySum(vector<int>& nums, int k) {
        long long maxSum = 0;
        long long sum = 0;
        int start = 0;
        int i = 0;

        unordered_map<int, int> mpp;

        while (i < nums.size()) {
            // cout  << i << endl;
            int key = nums[i];

            if (mpp.find(key) == mpp.end()) {
                mpp[key]++;
            } else {

                if(mpp[key] == 0){
                    mpp[key]++;
                }
                else{
                    mpp[nums[start]]--;
                    sum -= nums[start];
                    start++;
                    continue;
                }
                // i++;
                // continue;
            }

            sum += nums[i];
            if ((i - start + 1) == k) {
                // cout << start << " to  " << i << " = " << sum << endl;
                if(sum > maxSum){
                    maxSum = sum ;
                }
                mpp[nums[start]]--;
                sum -= nums[start];
                start++;
            }
            i++;
        }

        return maxSum;
    }
};

File: 2552-maximum-sum-of-distinct-subarrays-with-length-k/maximum-sum-of-distinct-subarrays-with-length-k.java
class Solution {
    public long maximumSubarraySum(int[] nums, int k) {
        HashMap<Integer,Integer> map = new HashMap<>();

        int i = 0 ;
        int j = 0 ;
        long sum = 0 ;
        long max = 0 ; 

        while(j < nums.length){

            int val = nums[j] ;
            sum += val;

            if(map.containsKey(val)){
                while(nums[i] != val){
                    map.remove(nums[i]);
                    sum -= nums[i];
                    i++;
                }

                map.remove(val);
                sum -= val ;
                i++;
            }

            else if(j-i+1 > k){
                map.remove(nums[i]);
                sum -= nums[i];
                i++;
            }

            map.put(val,j);



            if(j-i+1 == k){
                max = Math.max(max,sum);
            }
            
            j++;

        }

        return max ; 
    }
}